pkg/repository/v1_repository.go

```diff
@@ -121,194 +226,173 @@ func fnameWithVersion(fname string, version uint) string {
+	err = v1manifest.CheckExpiry(oldRoot.Expires)
 	if err != nil {
 		return errors.AddStack(err)
 	}
 
-	// Save the new trusted root.
-	err = r.local.SaveManifest(newTrusted, v1manifest.ManifestFilenameRoot)
+	// Save the new trusted root without a version number.
+	err = r.local.SaveManifest(newManifest, v1manifest.ManifestFilenameRoot)
 	if err != nil {
 		return errors.AddStack(err)
 	}
 
-	for _, m := range newRoots {
-		filename := fnameWithVersion(v1manifest.ManifestTypeRoot, m.Signed.Base().Version)
-		err = r.local.SaveManifest(m, filename)
-		if err != nil {
-			return errors.AddStack(err)
-		}
-	}
-
-	r.root = newTrusted.Signed.(*v1manifest.Root)
-
 	return nil
 }
 
-// Precondition: the index manifest actually requires updating, the root manifest has been updated if necessary.
-func (r *V1Repository) updateLocalIndex(length uint) (*v1manifest.Index, error) {
-	var root v1manifest.Root
-	_, err := r.local.LoadManifest(&root)
+// Precondition: the index manifest actually requires updating, snapshot manifest exists, and the root manifest has been updated if necessary.
+func (r *V1Repository) updateLocalIndex() error {
+	root, err := r.loadRoot()
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	var snapshot v1manifest.Snapshot
-	_, err = r.local.LoadManifest(&snapshot)
+	exists, err := r.local.LoadManifest(&snapshot)
 	if err != nil {
-		return nil, err
+		return err
+	}
+	if !exists {
+		return errors.New("No snapshot")
 	}
 
-	url, err := snapshot.VersionedURL(root.Roles[v1manifest.ManifestTypeIndex].URL)
+	url, fileVersion, err := snapshot.VersionedURL(root.Roles[v1manifest.ManifestTypeIndex].URL)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	var index v1manifest.Index
-	manifest, err := r.fetchManifest(url, &index, r.root, length)
+	manifest, err := r.fetchManifest(url, &index, fileVersion.Length)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	// Check version number against old manifest
 	var oldIndex v1manifest.Index
-	exists, err := r.local.LoadManifest(&oldIndex)
+	exists, err = r.local.LoadManifest(&oldIndex)
 	if exists {
 		if err != nil {
-			return nil, err
+			return err
 		}
 		if index.Version <= oldIndex.Version {
-			return nil, fmt.Errorf("index manifest has a version number <= the old manifest (%v, %v)", index.Version, oldIndex.Version)
+			return fmt.Errorf("index manifest has a version number <= the old manifest (%v, %v)", index.Version, oldIndex.Version)
 		}
 	}
 
-	err = r.local.SaveManifest(manifest, v1manifest.ManifestFilenameIndex)
-	if err != nil {
-		return nil, err
-	}
-
-	return &index, nil
+	return r.local.SaveManifest(manifest, v1manifest.ManifestFilenameIndex)
 }
 
-// Precondition: the snapshot manifest exists and is up to date
+// Precondition: the snapshot and index manifests exist and are up to date.
+// Returns nil if the component manifest is up to date.
 func (r *V1Repository) updateComponentManifest(id string) (*v1manifest.Component, error) {
```

Comment:
do you want to return the Component event the local one is up to date?
how can we get the v1manifest.Component?  will add some other kind of method?

and i think `UpdateComponents` still can not call this directly if it return `nil, nil` when the manifest is up to date

Reply:
The update functions leave the manifests on disk, so after calling `updateComponentManifest` and there is no error, then the guarantee is that the manifest on disk is the most recent (and validated). To access the manifest, use LocalManifests. You are correct, there is a bug in `UpdateManifests` such that we won't install a component if the manifest is up to date, I'll fix that...